src/detectors/builtin/sendInLoop.ts

```diff
@@ -0,0 +1,400 @@
+import { CompilationUnit } from "../../internals/ir";
+import { foldStatements } from "../../internals/tact";
+import { MistiTactWarning, Severity } from "../../internals/warnings";
+import { ASTDetector } from "../detector";
+import {
+  AstStatement,
+  AstStatementExpression,
+  AstExpression,
+  AstCondition,
+  AstStatementTry,
+  AstStatementTryCatch,
+  AstStatementWhile,
+  AstStatementRepeat,
+  AstStatementUntil,
+  AstStatementForEach,
+  AstFunctionDef,
+  AstAsmFunctionDef,
+} from "@tact-lang/compiler/dist/grammar/ast";
+
+/**
+ * A detector that identifies send function calls inside loops and loops that may cause out-of-gas exceptions.
+ *
+ * ## Why is it bad?
+ * **Send Functions Inside Loops:**
+ * - **Unexpected Behavior:** Multiple sends may occur, leading to unintended contract interactions.
+ * - **High Gas Consumption:** Loops with send calls can consume excessive gas, potentially causing out-of-gas exceptions.
+ * - **Reentrancy Vulnerabilities:** Sending messages within loops might open up reentrancy attack vectors.
+ *
+ * **Loops with Excessive Iterations:**
+ * - **Out-of-Gas Exceptions:** Loops with a large number of iterations may consume all available gas.
+ * - **Denial of Service:** Excessive gas consumption can be exploited to create DoS attacks.
+ *
+ * ## Example
+ * ```tact
+ * // Send function inside a loop
+ * while (i < limit) {
+ *   send(SendParameters{ to: recipient, ... });
+ * }
+ *
+ * // Loop with excessive iterations
+ * repeat (1_000_001) {
+ *   // ...
+ * }
+ * ```
+ *
+ * **Recommendation:** Refactor the code to avoid calling send functions inside loops and reduce the number of loop iterations.
+ */
+export class SendInLoop extends ASTDetector {
+  severity = Severity.HIGH;
+
+  private functionDefinitions: Map<string, AstFunctionDef | AstAsmFunctionDef> =
+    new Map();
+
+  async check(cu: CompilationUnit): Promise<MistiTactWarning[]> {
+    const warnings: MistiTactWarning[] = [];
+    this.collectFunctionDefinitions(cu);
+    this.collectWarnings(cu, warnings);
+    return warnings;
+  }
+
+  /**
+   * Collects function definitions from the compilation unit.
+   */
+  private collectFunctionDefinitions(cu: CompilationUnit): void {
+    for (const node of cu.ast.getProgramEntries()) {
+      if (node.kind === "function_def" || node.kind === "asm_function_def") {
+        const func = node as AstFunctionDef | AstAsmFunctionDef; // Updated type
+        const funcName = func.name.text; // Accessing 'name' is now valid
+        this.functionDefinitions.set(funcName, func);
+      }
+    }
+  }
+
+  /**
+   * Collects warnings by traversing the AST of the compilation unit.
+   */
+  private collectWarnings(
+    cu: CompilationUnit,
+    warnings: MistiTactWarning[],
+  ): void {
+    for (const node of cu.ast.getProgramEntries()) {
```

Comment:
nit: I would use `Array.from(cu.ast.getProgramEntries()).forEach(node => { ... })` here and in the similar cases. This makes code looks more consistent to the existent code base.